dataplane/src/pipeline/sample_nfs.rs

```diff
@@ -0,0 +1,96 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Open Network Fabric Authors
+
+use net::buffer::PacketBufferMut;
+use net::eth::mac::{DestinationMac, Mac};
+use net::headers::{TryEthMut, TryHeaders, TryIpv4Mut, TryIpv6Mut};
+use tracing::{debug, trace};
+
+use crate::packet::Packet;
+use crate::pipeline::NetworkFunction;
+
+/// Network function that uses [`debug!`] to print the parsed packet headers.
+pub struct InspectHeaders;
+
+impl<Buf: PacketBufferMut> NetworkFunction<Buf> for InspectHeaders {
+    fn process<'a, Input: Iterator<Item = Packet<Buf>> + 'a>(
+        &'a mut self,
+        input: Input,
+    ) -> impl Iterator<Item = Packet<Buf>> + 'a {
+        input.inspect(|packet| {
+            debug!("headers: {headers:?}", headers = packet.headers());
+        })
+    }
+}
+
+/// Network function that sets the destination mac address to the broadcast mac address.
+///
+/// The function has no effect if the packet is not an Ethernet packet.
+pub struct BroadcastMacs;
+
+impl<Buf: PacketBufferMut> NetworkFunction<Buf> for BroadcastMacs {
+    #[allow(clippy::unwrap_used, clippy::expect_used, clippy::panic)]
+    fn process<'a, Input: Iterator<Item = Packet<Buf>> + 'a>(
+        &'a mut self,
+        input: Input,
+    ) -> impl Iterator<Item = Packet<Buf>> + 'a {
+        input.map(|mut packet| {
+            match packet.try_eth_mut() {
+                None => {}
+                Some(mac) => {
+                    mac.set_destination(DestinationMac::new(Mac::BROADCAST).unwrap());
+                }
+            }
+            packet
+        })
+    }
+}
+
+/// Network function that decrements the TTL value of an IP packet.
+///
+/// The function has no effect if the packet is not an IP packet.
+/// If the TTL is 0, an error is logged using [`trace!`].
```

Comment:
nit: I would recommend adding these docs on the process method rather than at the type level.

Reply:
Really, what docs would apply to the entire trait then, since the whole purpose of the trait is to encapsulate that one function?

Reply:
I would leave the first sentence in the trait doc and the rest on the function.

The reasoning is mostly around IDE integration.

My IDE (and I assume most others) have solid integration with rustdoc and I think it is more common to mouse over / press a hotkey on the function call than on the trait to see the docs.
That is mostly because the trait is often not visible as text at the call site.

Reply:
Fixed.